ccl/multiregionccl: TestRegionAddDropWithConcurrentBackupOps failed

ccl/multiregionccl.TestRegionAddDropWithConcurrentBackupOps [failed](https://tanzanite.cluster.engflow.com/invocations/default/e1f109af-2318-4e0b-8241-09b07e7f8a41?testReportRun=5&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on release-24.1 @ [56ebe1cec09bcc38adbb2cbb4c16ec2e23dc228d](https://github.com/cockroachdb/cockroach/commits/56ebe1cec09bcc38adbb2cbb4c16ec2e23dc228d):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRegionAddDropWithConcurrentBackupOps (3m18s)
	TestRegionAddDropWithConcurrentBackupOps/add-region-succeed-backup-database (37s)

```
Stack: 

```
goroutine 1215739 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegionAddDropWithConcurrentBackupOps
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRegionAddDropWithConcurrentBackupOps502053290
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestRegionAddDropWithConcurrentBackupOps/add-region-fail-backup-database
=== RUN   TestRegionAddDropWithConcurrentBackupOps/add-region-succeed-backup-database
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionAddDropWithConcurrentBackupOps.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39625

ccl/multiregionccl.TestRegionAddDropWithConcurrentBackupOps [failed](https://tanzanite.cluster.engflow.com/invocations/default/9d273700-ca69-42e3-8485-b51d5ac38177?testReportRun=19&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on release-24.1 @ [dba532e47f9da689cc9103ee3f08b192998559b6](https://github.com/cockroachdb/cockroach/commits/dba532e47f9da689cc9103ee3f08b192998559b6):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRegionAddDropWithConcurrentBackupOps (3m26s)
	TestRegionAddDropWithConcurrentBackupOps/add-region-succeed-backup-database (54s)

```
Stack: 

```
goroutine 1131999 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRegionAddDropWithConcurrentBackupOps
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRegionAddDropWithConcurrentBackupOps4218355031
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestRegionAddDropWithConcurrentBackupOps/add-region-fail-backup-database
=== RUN   TestRegionAddDropWithConcurrentBackupOps/add-region-succeed-backup-database
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=19</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #126956 ccl/multiregionccl: TestRegionAddDropWithConcurrentBackupOps failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.2]
- #126928 ccl/multiregionccl: TestRegionAddDropWithConcurrentBackupOps failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.1.2-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionAddDropWithConcurrentBackupOps.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/ccl/multiregionccl/datadriven_test.go
+++ b/pkg/ccl/multiregionccl/datadriven_test.go
@@ -47,7 +47,31 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
-// TestMultiRegionDataDriven is a data-driven test to test various multi-region
+func TestMultiRegionDataDriven_global_tables(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	testMultiRegionDataDriven(t, "global_tables")
+}
+
+func TestMultiRegionDataDriven_regional_by_row(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	testMultiRegionDataDriven(t, "regional_by_row")
+}
+
+func TestMultiRegionDataDriven_regional_by_table(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	testMultiRegionDataDriven(t, "regional_by_table")
+}
+
+func TestMultiRegionDataDriven_secondary_region(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	testMultiRegionDataDriven(t, "secondary_region")
+}
+
+// testMultiRegionDataDriven is a data-driven test to test various multi-region
 // invariants at a high level. This is accomplished by allowing custom cluster
 // configurations when creating the test cluster and providing directives to
 // assert expectations in query traces.
@@ -112,10 +136,7 @@ import (
 //
 // "cleanup-cluster": destroys the cluster. Must be done before creating a new
 // cluster.
-func TestMultiRegionDataDriven(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
-
+func testMultiRegionDataDriven(t *testing.T, testPath string) {
 	// This test speeds up replication changes by proactively enqueuing replicas
 	// into various queues. This has the benefit of reducing the time taken after
 	// zone config changes, however the downside of added overhead. Disable the
@@ -124,7 +145,7 @@ func TestMultiRegionDataDriven(t *testing.T) {
 	skip.UnderRace(t, "flaky test")
 	skip.UnderDeadlock(t, "flaky test")
 	ctx := context.Background()
-	datadriven.Walk(t, datapathutils.TestDataPath(t), func(t *testing.T, path string) {
+	datadriven.Walk(t, datapathutils.TestDataPath(t, testPath), func(t *testing.T, path string) {
 		ds := datadrivenTestState{}
 		defer ds.cleanup(ctx)
 		var mu syncutil.Mutex

--- a/pkg/ccl/multiregionccl/region_test.go
+++ b/pkg/ccl/multiregionccl/region_test.go
@@ -847,7 +847,67 @@ CREATE DATABASE db WITH PRIMARY REGION "us-east1" REGIONS "us-east2" PLACEMENT R
 	}
 }
 
-// TestRegionAddDropEnclosingBackupOps tests adding/dropping regions
+func TestDropRegionFailWithConcurrentBackupOps(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testRegionAddDropWithConcurrentBackupOps(t, struct {
+		cmd                string
+		shouldSucceed      bool
+		expectedPartitions []string
+	}{
+		cmd:                `ALTER DATABASE db DROP REGION "us-east3"`,
+		shouldSucceed:      false,
+		expectedPartitions: []string{"us-east1", "us-east2", "us-east3"},
+	})
+}
+
+func TestDropRegionSucceedWithConcurrentBackupOps(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testRegionAddDropWithConcurrentBackupOps(t, struct {
+		cmd                string
+		shouldSucceed      bool
+		expectedPartitions []string
+	}{
+		cmd:                `ALTER DATABASE db DROP REGION "us-east3"`,
+		shouldSucceed:      true,
+		expectedPartitions: []string{"us-east1", "us-east2"},
+	})
+}
+
+func TestAddRegionFailWithConcurrentBackupOps(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testRegionAddDropWithConcurrentBackupOps(t, struct {
+		cmd                string
+		shouldSucceed      bool
+		expectedPartitions []string
+	}{
+		cmd:                `ALTER DATABASE db ADD REGION "us-east4"`,
+		shouldSucceed:      false,
+		expectedPartitions: []string{"us-east1", "us-east2", "us-east3"},
+	})
+}
+
+func TestAddRegionSucceedWithConcurrentBackupOps(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testRegionAddDropWithConcurrentBackupOps(t, struct {
+		cmd                string
+		shouldSucceed      bool
+		expectedPartitions []string
+	}{
+		cmd:                `ALTER DATABASE db ADD REGION "us-east4"`,
+		shouldSucceed:      true,
+		expectedPartitions: []string{"us-east1", "us-east2", "us-east3", "us-east4"},
+	})
+}
+
+// testRegionAddDropWithConcurrentBackupOps tests adding/dropping regions
 // (which may or may not succeed) with a concurrent backup operation
 // The sketch of the test is as follows:
 // - Client 1 performs an ALTER ADD / DROP REGION. Let the user txn commit.
@@ -858,43 +918,15 @@ CREATE DATABASE db WITH PRIMARY REGION "us-east1" REGIONS "us-east2" PLACEMENT R
 // - Restore the database, block in the schema changer.
 // - Fail or succeed the schema change job.
 // - Validate that the database and its tables look as expected.
-func TestRegionAddDropWithConcurrentBackupOps(t *testing.T) {
-	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
-
-	skip.UnderRace(t, "times out under race")
-
-	regionAlterCmds := []struct {
-		name               string
+func testRegionAddDropWithConcurrentBackupOps(
+	t *testing.T,
+	regionAlterCmd struct {
 		cmd                string
 		shouldSucceed      bool
 		expectedPartitions []string
-	}{
-		{
-			name:               "drop-region-fail",
-			cmd:                `ALTER DATABASE db DROP REGION "us-east3"`,
-			shouldSucceed:      false,
-			expectedPartitions: []string{"us-east1", "us-east2", "us-east3"},
-		},
-		{
-			name:               "drop-region-succeed",
-			cmd:                `ALTER DATABASE db DROP REGION "us-east3"`,
-			shouldSucceed:      true,
-			expectedPartitions: []string{"us-east1", "us-east2"},
-		},
-		{
-			name:               "add-region-fail",
-			cmd:                `ALTER DATABASE db ADD REGION "us-east4"`,
-			shouldSucceed:      false,
-			expectedPartitions: []string{"us-east1", "us-east2", "us-east3"},
-		},
-		{
-			name:               "add-region-succeed",
-			cmd:                `ALTER DATABASE db ADD REGION "us-east4"`,
-			shouldSucceed:      true,
-			expectedPartitions: []string{"us-east1", "us-east2", "us-east3", "us-east4"},
-		},
-	}
+	},
+) {
+	skip.UnderRace(t, "times out under race")
 
 	testCases := []struct {
 		name      string
@@ -909,158 +941,156 @@ func TestRegionAddDropWithConcurrentBackupOps(t *testing.T) {
 	}
 
 	for _, tc := range testCases {
-		for _, regionAlterCmd := range regionAlterCmds {
-			t.Run(regionAlterCmd.name+"-"+tc.name, func(t *testing.T) {
-				var mu syncutil.Mutex
-				typeChangeStarted := make(chan struct{})
-				typeChangeFinished := make(chan struct{})
-				backupOpFinished := make(chan struct{})
-				waitInTypeSchemaChangerDuringBackup := true
+		t.Run(tc.name, func(t *testing.T) {
+			var mu syncutil.Mutex
+			typeChangeStarted := make(chan struct{})
+			typeChangeFinished := make(chan struct{})
+			backupOpFinished := make(chan struct{})
+			waitInTypeSchemaChangerDuringBackup := true
 
-				backupKnobs := base.TestingKnobs{
-					SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-						RunBeforeEnumMemberPromotion: func(ctx context.Context) error {
-							mu.Lock()
-							defer mu.Unlock()
-							if waitInTypeSchemaChangerDuringBackup {
-								waitInTypeSchemaChangerDuringBackup = false
-								close(typeChangeStarted)
-								<-backupOpFinished
-							}
-							// Always return success here. The goal of this test isn't to
-							// fail during the backup, but to do so during the restore.
-							return nil
-						},
+			backupKnobs := base.TestingKnobs{
+				SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
+					RunBeforeEnumMemberPromotion: func(ctx context.Context) error {
+						mu.Lock()
+						defer mu.Unlock()
+						if waitInTypeSchemaChangerDuringBackup {
+							waitInTypeSchemaChangerDuringBackup = false
+							close(typeChangeStarted)
+							<-backupOpFinished
+						}
+						// Always return success here. The goal of this test isn't to
+						// fail during the backup, but to do so during the restore.
+						return nil
 					},
-					// Decrease the adopt loop interval so that retries happen quickly.
-					JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
-				}
+				},
+				// Decrease the adopt loop interval so that retries happen quickly.
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			}
 
-				tempExternalIODir, tempDirCleanup := testutils.TempDir(t)
-				defer tempDirCleanup()
+			tempExternalIODir, tempDirCleanup := testutils.TempDir(t)
+			defer tempDirCleanup()
 
-				_, sqlDBBackup, cleanupBackup := multiregionccltestutils.TestingCreateMultiRegionCluster(
-					t,
-					4, /* numServers */
-					backupKnobs,
-					multiregionccltestutils.WithBaseDirectory(tempExternalIODir),
-				)
-				defer cleanupBackup()
+			_, sqlDBBackup, cleanupBackup := multiregionccltestutils.TestingCreateMultiRegionCluster(
+				t,
+				4, /* numServers */
+				backupKnobs,
+				multiregionccltestutils.WithBaseDirectory(tempExternalIODir),
+			)
+			defer cleanupBackup()
 
-				_, err := sqlDBBackup.Exec(`
+			_, err := sqlDBBackup.Exec(`
 DROP DATABASE IF EXISTS db;
 CREATE DATABASE db WITH PRIMARY REGION "us-east1" REGIONS "us-east2", "us-east3";
 USE db;
 CREATE TABLE db.rbr(k INT PRIMARY KEY, v INT NOT NULL) LOCALITY REGIONAL BY ROW;
 INSERT INTO db.rbr VALUES (1,1),(2,2),(3,3);
 `)
-				require.NoError(t, err)
+			require.NoError(t, err)
 
-				go func(cmd string) {
-					defer func() {
-						close(typeChangeFinished)
-					}()
-					_, err := sqlDBBackup.Exec(cmd)
-					if err != nil {
-						t.Errorf("expected success, got %v when executing %s", err, cmd)
-					}
-				}(regionAlterCmd.cmd)
-
-				<-typeChangeStarted
-
-				_, err = sqlDBBackup.Exec(tc.backupOp)
-				close(backupOpFinished)
-				require.NoError(t, err)
-
-				<-typeChangeFinished
-
-				restoreKnobs := base.TestingKnobs{
-					SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
-						RunBeforeEnumMemberPromotion: func(context.Context) error {
-							mu.Lock()
-							defer mu.Unlock()
-							if !regionAlterCmd.shouldSucceed {
-								// Trigger a roll-back.
-								return jobs.MarkAsPermanentJobError(errors.New("nope"))
-							}
-							// Trod on.
-							return nil
-						},
-					},
-					// Decrease the adopt loop interval so that retries happen quickly.
-					JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			go func(cmd string) {
+				defer func() {
+					close(typeChangeFinished)
+				}()
+				_, err := sqlDBBackup.Exec(cmd)
+				if err != nil {
+					t.Errorf("expected success, got %v when executing %s", err, cmd)
 				}
+			}(regionAlterCmd.cmd)
 
-				// Start a new cluster (with new testing knobs) for restore.
-				_, sqlDBRestore, cleanupRestore := multiregionccltestutils.TestingCreateMultiRegionCluster(
-					t,
-					4, /* numServers */
-					restoreKnobs,
-					multiregionccltestutils.WithBaseDirectory(tempExternalIODir),
-				)
-				defer cleanupRestore()
+			<-typeChangeStarted
 
-				_, err = sqlDBRestore.Exec(tc.restoreOp)
-				require.NoError(t, err)
+			_, err = sqlDBBackup.Exec(tc.backupOp)
+			close(backupOpFinished)
+			require.NoError(t, err)
 
-				// First ensure that the data was restored correctly.
-				numRows := sqlDBRestore.QueryRow(`SELECT count(*) from db.rbr`)
-				require.NoError(t, numRows.Err())
-				var count int
-				err = numRows.Scan(&count)
-				require.NoError(t, err)
-				if count != 3 {
-					t.Logf("unexpected number of rows after restore: expected 3, found %d", count)
-				}
+			<-typeChangeFinished
 
-				// Now validate that the background job has completed and the
-				// regions are in the expected state.
-				testutils.SucceedsSoon(t, func() error {
-					dbRegions := make([]string, 0, len(regionAlterCmd.expectedPartitions))
-					rowsRegions, err := sqlDBRestore.Query("SELECT region FROM [SHOW REGIONS FROM DATABASE db]")
-					require.NoError(t, err)
-					defer func() {
-						require.NoError(t, rowsRegions.Close())
-					}()
-					for {
-						done := rowsRegions.Next()
-						if !done {
-							require.NoError(t, rowsRegions.Err())
-							break
+			restoreKnobs := base.TestingKnobs{
+				SQLTypeSchemaChanger: &sql.TypeSchemaChangerTestingKnobs{
+					RunBeforeEnumMemberPromotion: func(context.Context) error {
+						mu.Lock()
+						defer mu.Unlock()
+						if !regionAlterCmd.shouldSucceed {
+							// Trigger a roll-back.
+							return jobs.MarkAsPermanentJobError(errors.New("nope"))
 						}
-						var region string
-						err := rowsRegions.Scan(&region)
-						require.NoError(t, err)
-						dbRegions = append(dbRegions, region)
+						// Trod on.
+						return nil
+					},
+				},
+				// Decrease the adopt loop interval so that retries happen quickly.
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			}
+
+			// Start a new cluster (with new testing knobs) for restore.
+			_, sqlDBRestore, cleanupRestore := multiregionccltestutils.TestingCreateMultiRegionCluster(
+				t,
+				4, /* numServers */
+				restoreKnobs,
+				multiregionccltestutils.WithBaseDirectory(tempExternalIODir),
+			)
+			defer cleanupRestore()
+
+			_, err = sqlDBRestore.Exec(tc.restoreOp)
+			require.NoError(t, err)
+
+			// First ensure that the data was restored correctly.
+			numRows := sqlDBRestore.QueryRow(`SELECT count(*) from db.rbr`)
+			require.NoError(t, numRows.Err())
+			var count int
+			err = numRows.Scan(&count)
+			require.NoError(t, err)
+			if count != 3 {
+				t.Logf("unexpected number of rows after restore: expected 3, found %d", count)
+			}
+
+			// Now validate that the background job has completed and the
+			// regions are in the expected state.
+			testutils.SucceedsSoon(t, func() error {
+				dbRegions := make([]string, 0, len(regionAlterCmd.expectedPartitions))
+				rowsRegions, err := sqlDBRestore.Query("SELECT region FROM [SHOW REGIONS FROM DATABASE db]")
+				require.NoError(t, err)
+				defer func() {
+					require.NoError(t, rowsRegions.Close())
+				}()
+				for {
+					done := rowsRegions.Next()
+					if !done {
+						require.NoError(t, rowsRegions.Err())
+						break
 					}
-					if len(dbRegions) != len(regionAlterCmd.expectedPartitions) {
-						return errors.Newf("unexpected number of regions, expected: %v found %v",
+					var region string
+					err := rowsRegions.Scan(&region)
+					require.NoError(t, err)
+					dbRegions = append(dbRegions, region)
+				}
+				if len(dbRegions) != len(regionAlterCmd.expectedPartitions) {
+					return errors.Newf("unexpected number of regions, expected: %v found %v",
+						regionAlterCmd.expectedPartitions,
+						dbRegions,
+					)
+				}
+				for i, expectedRegion := range regionAlterCmd.expectedPartitions {
+					if expectedRegion != dbRegions[i] {
+						return errors.Newf("unexpected regions, expected: %v found %v",
 							regionAlterCmd.expectedPartitions,
 							dbRegions,
 						)
 					}
-					for i, expectedRegion := range regionAlterCmd.expectedPartitions {
-						if expectedRegion != dbRegions[i] {
-							return errors.Newf("unexpected regions, expected: %v found %v",
-								regionAlterCmd.expectedPartitions,
-								dbRegions,
-							)
-						}
-					}
-					return nil
-				})
-
-				// Finally, confirm that all of the tables were repartitioned
-				// correctly by the above ADD/DROP region job.
-				testutils.SucceedsSoon(t, func() error {
-					return multiregionccltestutils.TestingEnsureCorrectPartitioning(
-						sqlDBRestore,
-						"db",
-						"rbr",
-						[]string{"rbr@rbr_pkey"},
-					)
-				})
+				}
+				return nil
 			})
-		}
+
+			// Finally, confirm that all of the tables were repartitioned
+			// correctly by the above ADD/DROP region job.
+			testutils.SucceedsSoon(t, func() error {
+				return multiregionccltestutils.TestingEnsureCorrectPartitioning(
+					sqlDBRestore,
+					"db",
+					"rbr",
+					[]string{"rbr@rbr_pkey"},
+				)
+			})
+		})
 	}
 }

